src/repart/repart.c

```diff
@@ -2690,9 +2690,9 @@ static int context_copy_from_one(Context *context, const char *src) {
         if (r < 0)
                 return r;
 
-        r = fd_verify_regular(fd);
+        r = fd_verify_regular_or_block(fd);
         if (r < 0)
-                return log_error_errno(r, "%s is not a file: %m", src);
+                return log_error_errno(r, "%s is not a file nor a block device: %m", src);
```

Comment:
And then use `SYNTHETIC_ERRNO(ENOTTY)` here

Reply:
@poettering prefers propagating the error per https://github.com/systemd/systemd/pull/37209#discussion_r2054685037, so I'm back to returning `-EBADFD` in the case we're given something not a block device nor a regular file (same as what `stat_verify_regular` returns when not a regular file). That leads to the following message:

```bash
$ systemd-repart --definitions=$(mktemp -d) --copy-from=/dev/zero --empty=create --size=auto foo.raw
/dev/zero is not a file nor a block device: File descriptor in bad state
```